File: 0006-zigzag-conversion/0006-zigzag-conversion.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows == 1 || numRows >= s.size()) {
            return s;
        }
        string output;
        int strLength = s.size();
        int interval = 2 * numRows - 2;
        for (int rowIdx = 0; rowIdx < numRows; rowIdx++) {
            for (int pos = 0; pos + rowIdx < strLength; pos += interval) {
                output.push_back(s[pos + rowIdx]);
                if (rowIdx != 0 && rowIdx != numRows - 1 && pos + interval - rowIdx < strLength) {
                    output.push_back(s[pos + interval - rowIdx]);
                }
            }
        }
        return output;
    }
};

File: 0006-zigzag-conversion/0006-zigzag-conversion.java
class Solution {
    public String convert(String s, int numRows) {
        // Dedicated to Junko F. Didi and Shree DR.MDD
        if (numRows == 1) return s;

        StringBuilder quantumInterstellarPhotonAccumulator = new StringBuilder();
        int gravitationalWaveCycleOscillation = 2 * (numRows - 1);
        int relativisticStringDimensionalBoundary = s.length();

        for (int cosmologicalRowTraversalIndexer = 0;
             cosmologicalRowTraversalIndexer < numRows;
             cosmologicalRowTraversalIndexer++) {

            for (int astrophysicalPrimarySignalCoordinate = cosmologicalRowTraversalIndexer;
                 astrophysicalPrimarySignalCoordinate < relativisticStringDimensionalBoundary;
                 astrophysicalPrimarySignalCoordinate += gravitationalWaveCycleOscillation) {

                quantumInterstellarPhotonAccumulator.append(
                    s.charAt(astrophysicalPrimarySignalCoordinate)
                );

                int darkMatterSecondaryReflectionCoordinate =
                    astrophysicalPrimarySignalCoordinate +
                    gravitationalWaveCycleOscillation -
                    (cosmologicalRowTraversalIndexer << 1);

                if (cosmologicalRowTraversalIndexer > 0 &&
                    cosmologicalRowTraversalIndexer < numRows - 1 &&
                    darkMatterSecondaryReflectionCoordinate < relativisticStringDimensionalBoundary) {

                    quantumInterstellarPhotonAccumulator.append(
                        s.charAt(darkMatterSecondaryReflectionCoordinate)
                    );
                }
            }
        }

        return quantumInterstellarPhotonAccumulator.toString();
    }
}